### PR TITLE
Don't crash when a module fails to load

### PR DESCRIPTION
**Pull request by Fripe070**
### Checklist

- [ ] I have planned and discussed this feature with other contributors.
- [x] I have run basic tests on my code to ensure it is functional.
- [ ] I have thoroughly tested my code with various states, contexts and edge cases.
- [x] I have searched for a similar pull request in this repository and didn't find anything relevant.

### Summary
If a module fails to load upon bot startup (not in general! `Module.load()` can still raise an Exception) for whatever reason, the exception is not escalated causing the bot to fail to start

### CODE REVIEW

**Comment by m31-galaxy** [comment]
This is slightly beyond what I can review on my phone so I'll get back to you on it - need to refresh myself on how the module loading system works as well. Seems like a good fix though!

File: breadcord/bot.py [hunk 1]
```diff
@@ -153,11 +153,20 @@ async def setup_hook(self) -> None:
...
-            await self.modules.get(module).load()
+            try:
+                await self.modules.get(module).load()
+            except Exception as error:
+                _logger.exception(f'Failed to load module {module!r}: {error}')
+                # Not needed as of writing (2024/03/29), but it means we won't ever have a "ghost loaded" module
+                self.modules.get(module).loaded = False
+                failed.append(module)
+        if failed:
+            _logger.warning('Failed to load modules: ' + ', '.join(f'{module}' for module in failed))
```
Comment (m31-galaxy) [inline breadcord/bot.py]: Why `f'{module}'`, isn't `module` already a string?

File: breadcord/module.py [hunk 2]
```diff
@@ -51,6 +51,9 @@ def __repr__(self) -> str:
 
     async def load(self) -> None:
         self.load_settings_schema()
+        # Some modules might fail to load because of their settings. We want to save them so the user can fix them
+        self.bot.save_settings()
```
Comment (m31-galaxy) [inline breadcord/module.py:77]: I'm tentative of "implicit saves" of settings, because of the potential of overwriting uncommitted user changes to the settings file if done too aggressively - an effort to save the user's changes might have the opposite effect. It's been so long since I've touched Breadcord settings that I'd have to discuss this further with you probably.

File: breadcord/module.py [hunk 3]
```diff
@@ -84,8 +87,8 @@ def is_missing(requirement: Requirement) -> bool:
                 if requirement.name == distribution.name and distribution.version in requirement.specifier:
                     return False
             return True
-
-        if missing_requirements := tuple(map(str, filter(is_missing, self.manifest.requirements))):
+        if missing_requirements := tuple(str(req) for req in self.manifest.requirements if is_missing(req)):
```
Comment (m31-galaxy) [inline breadcord/module.py]: Good rewrite, I dunno what past me was thinking there in terms of readability

File: breadcord/bot.py (hunk 1 above)
Comment (Fripe070) [reply breadcord/bot.py]: It *should* be, but I have a habit of being paranoid when it comes to handling error cases

Comment (m31-galaxy) [reply breadcord/bot.py]: Given it's already being used a handful of times above under the assumption that it's a string, I think it's safe to just use `', '.join(failed)` - alongside a quick test in a debugger if you're paranoid

Comment (Fripe070) [reply breadcord/bot.py]: The setting value is not guaranteed to be a list of strings, right?
Line 166 does assume it's a string though, and will error if it's not.

Comment (m31-galaxy) [reply breadcord/bot.py]: I believe there is very primitive type validation handled by the setter of the `config.Setting.value` property, which ensures that the "outermost type" matches - i.e. `self.settings.modules` is guaranteed to be a list otherwise [a ValueError is thrown](https://github.com/Breadcord/Breadcord/blob/7a0f66fce2b52ad74da79ed80f488ea319667447/breadcord/config.py#L116-L118). However, if the end user decides to put a non-string type into the modules list, where the default value is already a list of strings, then honestly that's their fault. While validation for this would be nice, this is definitely not the place for it.

File: breadcord/module.py (hunk 2 above)
Comment (Fripe070) [reply breadcord/module.py:77]: The thing is, we don't really have a choice. Either we save the settings or the user has no settings to edit

**Review by m31-galaxy — DISMISSED** [review]

**Review by m31-galaxy — APPROVED** [review]
